<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Richard Judson" />


<title>httrpathway-vignette</title>

<script>// Hide empty <a> tag within highlighted CodeBlock for screen reader accessibility (see https://github.com/jgm/pandoc/issues/6352#issuecomment-626106786) -->
// v0.0.1
// Written by JooYoung Seo (jooyoung@psu.edu) and Atsushi Yasumoto on June 1st, 2020.

document.addEventListener('DOMContentLoaded', function() {
  const codeList = document.getElementsByClassName("sourceCode");
  for (var i = 0; i < codeList.length; i++) {
    var linkList = codeList[i].getElementsByTagName('a');
    for (var j = 0; j < linkList.length; j++) {
      if (linkList[j].innerHTML === "") {
        linkList[j].setAttribute('aria-hidden', 'true');
      }
    }
  }
});
</script>

<style type="text/css">
  code{white-space: pre-wrap;}
  span.smallcaps{font-variant: small-caps;}
  span.underline{text-decoration: underline;}
  div.column{display: inline-block; vertical-align: top; width: 50%;}
  div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
  ul.task-list{list-style: none;}
    </style>


<style type="text/css">
  code {
    white-space: pre;
  }
  .sourceCode {
    overflow: visible;
  }
</style>
<style type="text/css" data-origin="pandoc">
code.sourceCode > span { display: inline-block; line-height: 1.25; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */

</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    for (var j = 0; j < rules.length; j++) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") continue;
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') continue;
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>




<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">httrpathway-vignette</h1>
<h4 class="author">Richard Judson</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1"></a><span class="co">#library(httrpathway)</span></span></code></pre></div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>“httrpathway” is a set of functions for calculating concentration-response data for high-throughput transcriptomics data (HTTr). To-date, this package has been used to process TempO-Seq data, but in principle, and RNASeq data that is properly preprocessed could be used. Input data is assume to have been created in the form of log2-fold change values (l2fc) for each probe (or gene) for each chemical sample x concentration pair. We currently use DESeq2 to convert normalized count data to this format. Concentration-response modeling is done in two ways. The first analyzes the data at the level of individual genes, and the second at the level of signatures (genesets or pathways). This package inputs and outputs large files containing the data, so assumes a somewhat complicated directory structure. This vignette describes the directory structure, the process need to create the input files, and the process of running and analyzing gene and signature-level concentration-response modeling.</p>
</div>
<div id="directory-structure" class="section level2">
<h2>Directory Structure</h2>
<p>The package and all of the directories listed here are assumed to reside in a parent directory (called here “parent”). In parent are the following directories/ Some of these will have subdirectories for individual cell types or data sets. Before running the code, these main directories need to be created.</p>
<ul>
<li>parent/httrpathway - the R package</li>
<li>parent/input/signatures - contains the signature data</li>
<li>parent/input/fcdata - contains the input l2fc data</li>
<li>parent/output/gene_conc_resp_summary - concentration-response results at the gene level</li>
<li>parent/output/gene_conc_resp_plots - concentration-response plots at the gene level</li>
<li>parent/output/signature_cutoff - signature cutoffs (intermediate results)</li>
<li>parent/output/signature_conc_resp_summary - concentration-response results at the signature level</li>
<li>parent/output/signature_conc_resp_plots - concentration-response plots at the signature level. One folder per data set (auto-created)</li>
<li>parent/output/signature_score_summary - signature scores for each samples, conenttration, signature - intermediate results</li>
<li>parent/output/super_target_boxplot - super-target plots to summarize signature data at hte chemical level.</li>
</ul>
</div>
<div id="signatures" class="section level2">
<h2>Signatures</h2>
<p>There are 3 main files in the parent/input/signatures directory</p>
<ul>
<li>signatureDB.RData - this is a data.frame with the key columns from signature catalog, plus a column gene.list where the gene symbols for the signature are given in pipe-delimited form (e.g. gene1|gene2|gene3)</li>
<li>signatureDB_genelists.RData - this is the signatures as a list of lists. Each element in the list is a named list (the signature names) followed by a list of the contained genes. To add signatures to either of these files, one need to write a special piece of code. Note that thes files names are required by the code and should not be changed.</li>
<li>signatureDB_master_catalog {date}.xlsx - This is a catalog of the signatures that provides a variety of annotation. The name of this file is user-defined, and is an input variable in certain functions. The catalog is described next.</li>
</ul>
<p>Signature are lists of genes that come from in a variety of types. Some are genes perturbed by specific chemicals or other molecular perturbation, some are genes perturbed in certain disease or conditions and some are genes involved in specific molecular pathways. For some classes of signatures, there are separate sets of genes that are expected to increase in expression and others that are expected to decrease. Therefore signatures cna be either directional of non-directional. All signatures have a name, and a parent. For nondirectional signatures, these are the same, but for directional signatures, the individual signature names will be “parent_up” and “parent_dn”. The final score for a directional signature pair is score(up)-score(dn), and this is assigned to the parent. Key fields in the catalog are:</p>
<ul>
<li>signature - the name of the signature</li>
<li>parent - the name of the parent. In the outputs, this is the name used</li>
<li>source</li>
<li>subsource</li>
<li>type - directional or nondirectional</li>
<li>direction - up, dn or nondirectional</li>
<li>ngene - number of genes in the signature. Note that fewer than this number may be used becasue some of these genes may not occur in the input data set (expression was too low)</li>
<li>description - a description of the signature</li>
<li>super_target - a standard target annotation, often a gene, gene family, disease or biological process</li>
<li>effect_direction - expected direction, mainly useful in the case of signatures arising from agonists, antagonists, inhibitors, etc.</li>
<li>target_class - a high-level descriptive term for the target class (e.g. enzyme, transporter)</li>
<li>super_target_level - the level of biological organization of the super_target (chemical, gene, cell, tissue, organ)</li>
<li>include0 - a flag to set if the signature is not to be used (0 to exclude, 1 to include)</li>
<li>signature set names (pilot_small, screen_large, …) - When running calculations, one rarely wants to use all available signatures, so a column is used that designates the signatures to be used (yes=1, no=0). The column name is then the variable <em>sigset</em> (signature set) which is an input to several functions. Note that pairs of directional signatures need to be included or excluded together.</li>
</ul>
<p><strong>Key note about signature sizes</strong>: A data set (variable name <em>dataset</em>) will not contain all genes in the genome - only those with above a certain level of expression, maybe 5 counts per million in the majority of samples. When calculations are being run, the signature gene lists are subsetted to those in the data set. Calculations are then only run with signatures containing at least 10 genes after this subsetting. The code is not very smart, so these newly shrunk signatures need to be excluded and the code restarted. A warning message about these bad signatures is given and the code halts. Edit the signature catalog, excluding the signatures from the current <em>sigset</em> and restart. This check happens early. If one of a directional pair is excluded, both need to be excluded.</p>
</div>
<div id="creating-input-l2fc-files" class="section level2">
<h2>Creating input l2fc files</h2>
<p>For each data set to be run (variable <em>dataset</em>), 2 input files are needed, the chemical dictionary and the l2fc matrix. A dataset comprises the data for a group of chemicals run in a particular cell type. These are named CHEM_DICT_{dataset}.RData and FCMAT2_{dataset}.RData. CHEM_DICT has one row per chemcial sample x concentration combination. Key variables are the chemical name, the DSSTox Substance_id (variable name <em>dtxsid</em>), the concentration, the concentraiton index and the sample_key (a concatenation of these other variables). FCMAT2 is a data.frame with one row per row in CHEM_DICT, with rownames being the sample key and columns being genes (or probes). In our implementation, we create these files from the DESeq2 output by running the following functions</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1"></a><span class="kw">buildFCMAT1.fromDB</span>(<span class="dt">dataset=</span>dataset,<span class="dt">dir=</span>{location of DESeq2 file},<span class="dt">infile=</span>{name of DESEQ2 file}, <span class="dt">do.load=</span>T)</span>
<span id="cb2-2"><a href="#cb2-2"></a><span class="kw">buildFCMAT2.fromDB</span>(<span class="dt">dataset=</span>dataset,<span class="dt">time=</span>{incubation time <span class="cf">in</span> hours},<span class="dt">media=</span>{cell media},<span class="dt">method=</span><span class="st">&quot;gene&quot;</span>,<span class="dt">do.read=</span>T) </span></code></pre></div>
<p>The resulting CHEM_DICT and FCMAT2 files will be placed in parent/input/fcdata. These functions currently produce a number of other files (for instance the standard error matrix in the same format as the l2fc matrix) that might prove useful.</p>
</div>
<div id="running-concentration-response-modeling---signatures" class="section level2">
<h2>Running Concentration Response Modeling - Signatures</h2>
<p>To run the concentration response process, use the function <em>driver.R</em>, shown here with standard settings. The core methods for performing concentration-response modeling is the <em>tcplfit2</em> package on <a href="https://cran.r-project.org/web/packages/tcplfit2/index.html">CRAN</a>, so one should be familiar with that to help understand the following.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1"></a><span class="kw">driver</span>(<span class="dt">dataset=</span>dataset,</span>
<span id="cb3-2"><a href="#cb3-2"></a>       <span class="dt">sigcatalog=</span>{name of the signature catalog file, without the .xlsx extension},</span>
<span id="cb3-3"><a href="#cb3-3"></a>       <span class="dt">sigset=</span>sigset,</span>
<span id="cb3-4"><a href="#cb3-4"></a>       <span class="dt">mc.cores=</span>{<span class="dv">1</span> <span class="cf">for</span> a serial machine, or <span class="op">&gt;=</span><span class="dv">1</span> on a paralell system},</span>
<span id="cb3-5"><a href="#cb3-5"></a>       <span class="dt">nlowconc=</span><span class="dv">2</span>,</span>
<span id="cb3-6"><a href="#cb3-6"></a>       <span class="dt">hccut=</span><span class="fl">0.9</span>,</span>
<span id="cb3-7"><a href="#cb3-7"></a>       <span class="dt">tccut=</span><span class="dv">1</span>,</span>
<span id="cb3-8"><a href="#cb3-8"></a>       <span class="dt">plotrange=</span><span class="kw">c</span>(<span class="fl">0.0001</span>,<span class="dv">100</span>),</span>
<span id="cb3-9"><a href="#cb3-9"></a>       <span class="dt">method=</span><span class="st">&quot;gsea&quot;</span>,</span>
<span id="cb3-10"><a href="#cb3-10"></a>       <span class="dt">celltype=</span>{name of the cell type},</span>
<span id="cb3-11"><a href="#cb3-11"></a>       <span class="dt">do.conc.resp=</span>T,</span>
<span id="cb3-12"><a href="#cb3-12"></a>       <span class="dt">do.scr.plots=</span>T,</span>
<span id="cb3-13"><a href="#cb3-13"></a>       <span class="dt">do.supertarget.boxplot=</span>T)</span></code></pre></div>
<p>A few variables might be changed as follows</p>
<ul>
<li>mc.cores - Several sections of the code will run on a parallel machine, so you can add the number of cores. We have run up to 40 cores</li>
<li>nlowconc - In doing concentration-response modeling, a key parameter is <em>cutoff</em>, which is the 95% confidence interval around the assumed background. For a signature-chemical combination to be called active, at least one point needs to be above this (or below <em>-cutoff</em>). The standard way to calculate this cutoff is to take (for each signature), the score for the lowest <em>nlowconc</em> concentrations for all samples/chemicals in the data set and calculate the 95th percentile from this. This assumes that for most signatures and most chemicals, there is little bioactivity at the few lowest concentrations. We have looked at how few chemical one needs to get a reasonable cutoff and have found that for as few as 10, the cutoffs are close to what one gets from 1000 chemicals. For very small or very active data sets, some other way to determine cutoffs must be used. In that case, a cutoff file can be specified and this option is not used (see the full documentation for <em>driver.R</em>)</li>
<li>hccut - The output of the concentration-response process is a vector for each chemical sample / signature pair, including the variable <em>hitcall</em>, which is in the range of 0-1. A high value indicates high confidence that there is activity. For some downstream analyses, <em>hitcall</em> is binarized with values <em>&lt;hccut</em> set to inactive. The value of 0.9 is our current default.</li>
<li>tccut - Another output variable from the fitting is <em>top_over_cutoff</em>, a kind of Z-score. In some cases, it may be useful to filter results where <em>hitcall</em> ~1, but where <em>top_over_cutoff</em> is above but close to 1. Currently we do not do this.</li>
<li>method - This variable is designates what signature scoring method to use and it is currently either gsea (Gene Set Enrichment Analysis) or fc (fold change).We currently use only gsea, but are evaluating other scoring functions to better distinguish signal from noise.</li>
<li>do.X - These flags let you run jsut specific parts of the code (calculate concentration-response, produce concentration-response plots or produce the super_target boxplots to summarize all the results in one plot per chemical sample)</li>
</ul>
</div>
<div id="running-concentration-response-modeling---genes" class="section level2">
<h2>Running Concentration Response Modeling - Genes</h2>
<p>To run concentration-response modeling at hte level of individual genes, use the following function</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1"></a><span class="kw">geneConcResp</span>(<span class="dt">dataset=</span>dataset,</span>
<span id="cb4-2"><a href="#cb4-2"></a>             <span class="dt">mc.cores=</span><span class="dv">1</span>,</span>
<span id="cb4-3"><a href="#cb4-3"></a>             <span class="dt">genefile=</span><span class="ot">NULL</span>) {</span></code></pre></div>
<p>This is much simpler than the call (and code) for running signature concentration-response modeling. The only new variable is <em>genefile</em>. If one wants to only run the code for a subset of genes, provide an Excel file (with a relative path) containing the gene list, one gene per line.</p>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
